pkg/innerring/processors/netmap/process_epoch.go

```diff
@@ -65,14 +68,64 @@ func (np *Processor) processNewEpoch(ev netmapEvent.NewEpoch) {
 		}
 	}
 
-	np.netmapSnapshot.update(*networkMap, epoch)
+	if np.netmapSnapshot.update(*networkMap, epoch) {
+		l.Debug("updating placements in Container contract...")
+		err = np.updatePlacementInContract(*networkMap, l)
+		if err != nil {
+			l.Error("can't update placements in Container contract", zap.Error(err))
+		} else {
+			l.Debug("updated placements in Container contract")
+		}
+	}
 	np.handleCleanupTick(netmapCleanupTick{epoch: epoch, txHash: ev.TxHash()})
 	np.handleNewAudit(audit.NewAuditStartEvent(epoch))
 	np.handleAuditSettlements(settlement.NewAuditEvent(epoch))
 	np.handleAlphabetSync(governance.NewSyncEvent(ev.TxHash()))
 	np.handleNotaryDeposit(ev)
 }
 
+func (np *Processor) updatePlacementInContract(nm netmap.NetMap, l *zap.Logger) error {
+	// TODO: https://github.com/nspcc-dev/neofs-node/issues/3045
+	cids, err := np.containerWrp.List(nil)
+	if err != nil {
+		return fmt.Errorf("can't get containers list: %w", err)
+	}
+
+	for _, cID := range cids {
```

Comment:
@roman-khimov, i think this is the only thing that we may discuss: IMO if netmap is changed, many containers change their placement and it is easier to recalculate them all once in a single epoch (and not touch them for many-many epochs) than to cache the whole map from the previous epoch OR to fetch it, calculate and compare every vector in every container to see if anything has changed

Reply:
Works for now.